app/src/main/kotlin/com/wire/android/feature/conversation/content/domain/MessageSender.kt

```diff
@@ -0,0 +1,76 @@
+package com.wire.android.feature.conversation.content.domain
+
+import com.wire.android.core.async.DispatcherProvider
+import com.wire.android.core.exception.Failure
+import com.wire.android.core.exception.NetworkConnection
+import com.wire.android.core.exception.Unauthorized
+import com.wire.android.core.functional.Either
+import com.wire.android.core.functional.suspending
+import com.wire.android.feature.conversation.content.Message
+import com.wire.android.feature.conversation.content.MessageRepository
+import com.wire.android.feature.conversation.content.SendMessageFailure
+import com.wire.android.feature.messaging.ChatMessageEnvelope
+import com.wire.android.shared.session.SessionRepository
+import kotlinx.coroutines.withContext
+
+class MessageSender(
+    private val messageRepository: MessageRepository,
+    private val sessionRepository: SessionRepository,
+    private val messageSendFailureHandler: MessageSendFailureHandler,
+    private val outgoingMessageRecipientsRetriever: OutgoingMessageRecipientsRetriever,
+    private val messageEnvelopeCreator: MessageEnvelopeCreator,
+    dispatcherProvider: DispatcherProvider
+) {
+
+    private val singleThreadedContext = dispatcherProvider.newSingleThreadedDispatcher(WORK_POOL_NAME)
```

Comment:
For curiosity, why are you using newSingleThreadedDispatcher?

Reply:
To make sure messages are sent in order one at a time and no duplicated work is done.

If, by any reason we have multiple calls to send a message, it may be that the order is not preserved or that we fetch multiple preKeys for the same device and try to establish multiple sessinos for it without any need.

The `singleThreadDispatcher` will make sure we process one by one.

Reply:
Thank you 👌